engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java

```diff
@@ -214,15 +217,24 @@ public boolean isSuitableForEvent(final ExecutableEndEvent element) {
 
     @Override
     public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
-
-      final var activated = stateTransitionBehavior.transitionToActivated(activating);
-      final boolean canBeCompleted =
-          eventPublicationBehavior.throwEscalationEvent(
-              element.getId(), element.getEscalation(), activated);
-
-      if (canBeCompleted) {
-        stateTransitionBehavior.completeElement(activated);
-      }
+      expressionProcessor
+          .evaluateStringExpressionAsDirectBuffer(
+              element.getEscalation().getEscalationCodeExpression(),
+              activating.getElementInstanceKey())
+          .ifRightOrLeft(
+              escalationCode -> {
+                final var escalation = element.getEscalation();
+                escalation.setEscalationCode(escalationCode);
```

Comment:
❌ We shouldn't change the model by setting the `escalationCode` here. We should move the evaluation of this expression to the `BpmnEventPublicationBehavior.throwEscalationEvent` method. This also removes the duplication from this processor and the intermediate throw event processor.

Reply:
❓ I don't think we need to move it into the `BpmnEventPublicationBehavior.throwEscalationEvent` method, when we fail to evaluate the escalation expression, we should create an incident. 

----

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L139-L146
So I think we can change the `throwEscalationEvent` method signature to this:
```java
public boolean throwEscalationEvent(
      final DirectBuffer throwElementId,
      final DirectBuffer escalationCode,
      final BpmnElementContext context) {
      
      // do something.
      }
```

@remcowesterhoud. WDYT?

Reply:
❓ I don't think we need to move it into the `BpmnEventPublicationBehavior.throwEscalationEvent` method, and the expression should be evaluated first. If the expression evaluation fails, we will create a incident instead of evaluating the expression after the current element is activated.

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java#L218-L221


@remcowesterhoud. WDYT?

Reply:
https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L139-L146

---

BTW. We can modify the signature of the `throwEscalationEvent` method to avoid changing the model.
```java
public Either<Failure, Boolean> throwEscalationEvent(
      final DirectBuffer throwElementId,
      final DirectBuffer escalationCode,
      final BpmnElementContext context) {

      // do something.
}
```

Reply:
I was thinking we can have the `BpmnEventPublicationBehavior` return an `Either<Failure, Boolean>`. This would involve adding a method here like this which can take care of the evaluation for us:

```java
  public Either<Failure, Boolean> throwEscalationEvent(
      final DirectBuffer throwElementId,
      final ExecutableEscalation escalation,
      final BpmnElementContext context) {
    final var expressionEvaluationResult =
        expressionProcessor.evaluateStringExpressionAsDirectBuffer(
            escalation.getEscalationCodeExpression(), context.getElementInstanceKey());

    if (expressionEvaluationResult.isLeft()) {
      return Either.left(expressionEvaluationResult.getLeft());
    } else {
      final DirectBuffer escalationCode = expressionEvaluationResult.get();
      return Either.right(throwEscalationEvent(throwElementId, escalationCode, context));
    }
  }
```

Note that this calls the current (now private scopes) `throwEscalationEvent()` method. We just need to change the `escalation` parameter to `escalationCode`.

This way we extract this logic from our processors. Making them a bit simpler:

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      final BpmnElementContext activated =
          stateTransitionBehavior.transitionToActivated(activating);
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activated)
          .ifRightOrLeft(
              canBeCompleted -> {
                if (canBeCompleted) {
                  stateTransitionBehavior.completeElement(activated);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activated));
    }
```

Reply:
Thanks. @remcowesterhoud. ❤️ 

That's how it was originally planned. I was worried that the current element would be activated multiple times during the resolve incident.

Reply:
Looking at other parts of the code we do the same thing, so this shouldn't be a problem

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      final var activated = stateTransitionBehavior.transitionToActivated(activating);
      final var completing = stateTransitionBehavior.transitionToCompleting(activated);
      stateTransitionBehavior
          .transitionToCompleted(element, completing)
          .ifRightOrLeft(
              completed -> {
                // terminate all other element instances of the same flow scope
                final var flowScopeContext = stateBehavior.getFlowScopeContext(completed);
                stateTransitionBehavior.terminateChildInstances(flowScopeContext);
              },
              failure -> incidentBehavior.createIncident(failure, completing));
    }
```

Reply:
> Looking at other parts of the code we do the same thing, so this shouldn't be a problem

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/ResolveIncidentProcessor.java#L135-L145

---
```
java.lang.IllegalStateException: Expected to continue processing after incident 2251799813685256 resolved, but failed command not found
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.lambda$attemptToContinueProcessProcessing$1(ResolveIncidentProcessor.java:113) ~[classes/:?]
	at io.camunda.zeebe.util.Either$Left.ifRightOrLeft(Either.java:476) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.attemptToContinueProcessProcessing(ResolveIncidentProcessor.java:100) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.processRecord(ResolveIncidentProcessor.java:76) ~[classes/:?]
	at io.camunda.zeebe.engine.Engine.process(Engine.java:128) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$3(ProcessingStateMachine.java:264) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:260) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:209) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:185) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[classes/:?]
Caused by: java.lang.IllegalStateException: Expected incident to refer to element in state ELEMENT_ACTIVATING or ELEMENT_COMPLETING, but element is in state ELEMENT_ACTIVATED
```

@remcowesterhoud. Seems like you're missing something? 🤔

Reply:
Interesting, I wasn't aware of this. Sorry! I suppose we should wait with transitioning to activated in the processor, which makes sense now that I think about it.

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activating)
          .ifRightOrLeft(
              canBeCompleted -> {
                final BpmnElementContext activated =
                    stateTransitionBehavior.transitionToActivated(activating);
                if (canBeCompleted) {
                  stateTransitionBehavior.completeElement(activated);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));
    }
```

Reply:
By the way, this previous example is not entirely correct. We should use `stateTransitionBehavior.transitionToCompleted(element, completing)` as we do in the other processors.

Reply:
❓ @remcowesterhoud If we modify it like this:

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activating)
          .ifRightOrLeft(
              canBeCompleted -> {
                final BpmnElementContext activated =
                    stateTransitionBehavior.transitionToActivated(activating);
                if (canBeCompleted) {
                  final var completing = stateTransitionBehavior.transitionToCompleting(activated);
                  stateTransitionBehavior.completeElement(completing);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));
    }
```

The whole semantic becomes to activate `escalation catch event` first, and then `escalation throw event`, do you think this is as expected?

Reply:
Personally I like that. It means the main functionality happens between the `ACTIVATING` and the `ACTIVATED` state of the element instance. However, it does differ from the error end event. So I'll gather some opinions on this and get back to you.

Reply:
Hi @lzgabel, I have an answer surrounding this topic. Although functionality-wise it does not matter we want to keep it aligned with the error event. As a result we will have to pull the finding of the catch event out of the throw method again. Sorry for the extra work here.

So the processor would look somewhat like this:
- Find the escalation catch event. I suppose this will return an `Either<Failure, Optional<CatchEventTuple>>`
- On success:
    - Transition to activated
    - If catch event found, throw the catch event
    - Transition to completing (if needs to be completed)
    - Transition to completed (if needs to be completed)
- On failure:
    - Create incident
    
I would propose to change the `findEscalationCatchEvent` method to take an `ExecutableEscalation` instead of an `escalationCode`. We can do the expression evaluation in this method so it is still reused by both processors.
The `throwEscalationEvent` will not need the `ExecutableEscalation` anymore and can take a `CatchEventTuple` instead.

I hope this clears things up 🙂 Thanks for keeping me sharp!

Reply:
Just in case it's not clear, as I may have written this down somewhat confusing. This flow is not all in the `onActivate` method. The transitioning to completed should be part of the `onComplete()` method

Reply:
> The throwEscalationEvent will not need the ExecutableEscalation anymore and can take a CatchEventTuple instead.

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L148-L151

❓ Hi @remcowesterhoud. When the catch event is not found, if we do not pass `ExecutableEscalation` here, we will not be able to obtain `escalationCode` at this time. Do you have other ideas?

Reply:
I think we should extract this from this method and pull into a separate method. So in the processor it'd do:

- Catch event is present?
    - Yes? Call the behavior to trigger the catch event and write the escalated record
    - No? Call the behavior to write the not escalated record

I'm a bit stuck on the naming for these 2 methods at the moment. I guess it could be as simple as `throwEscalationEvent(throwElementId, CatchEventTuple, BpmnElementContext)` and `writeNotEscalatedEvent(throwElementId, errorCode, BpmnElementContext)`

Do you think this will work?

Reply:
Hi @remcowesterhoud. I think we will still face the problem of evaluating expressions multiple times.

```java

eventPublicationBehavior
          .findEscalationCatchEvent(escalation, activating)
          .ifRightOrLeft(
              catchEventTuple -> {
                final var throwElementId = element.getId();
                if (catchEventTuple.isPresent()) { // catch event is found
                  eventPublicationBehavior.throwEscalationEvent(
                      throwElementId, catchEventTuple.get(), activating);
                } else { // catch event is not found
                  final Optional<DirectBuffer> escalationCodeOptional = element.getEscalation()
                      .getEscalationCode();
                  DirectBuffer escalationCode;
                  if (escalationCodeOptional.isEmpty()) { 
                    final Either<Failure, DirectBuffer> expressionEvaluationResult = expressionProcessor.evaluateStringExpressionAsDirectBuffer(
                        escalation.getEscalationCodeExpression(),
                        activating.getElementInstanceKey());
                    escalationCode = expressionEvaluationResult.get();
                  } else {
                    escalationCode = escalationCodeOptional.get();
                  }
                  eventPublicationBehavior.writeNotEscalatedEvent(
                      throwElementId,
                      escalationCode,
                      activating);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));

```

Reply:
We've come full circle 😄 I'll have to accept the we will do the evaluation in the processor and have it duplicated in the other processor. I don't like evaluating it twice so let's avoid this. And since we still want the sequence of the record to be as expected I currently don't see a way around this.
This means the implementation here would be very similar to the error end event 😛

Reply:
So, let's keep the status quo? Or looking for a different solution? 🤔

Reply:
Yes, let's keep the status quo. Implement it similar to the error end event. We can always improve upon it later if we don't like it.